pkg/backup/actions/csi/pvc_action.go

```diff
@@ -570,3 +581,485 @@ func NewPvcBackupItemAction(f client.Factory) plugincommon.HandlerInitializer {
+		return val, nil
+	}
+
+	// 3. Fallback to label-based default
+	vgsClassList := &volumegroupsnapshotv1beta1.VolumeGroupSnapshotClassList{}
+	if err := p.crClient.List(context.TODO(), vgsClassList); err != nil {
+		return "", errors.Wrap(err, "failed to list VolumeGroupSnapshotClasses")
+	}
+
+	var matched []string
+	for _, class := range vgsClassList.Items {
+		if class.Driver != driver {
+			continue
+		}
+		if val, ok := class.Labels[velerov1api.VolumeGroupSnapshotClassDefaultLabel]; ok && val == "true" {
+			matched = append(matched, class.Name)
+		}
+	}
+
+	if len(matched) == 1 {
+		return matched[0], nil
+	} else if len(matched) == 0 {
+		return "", errors.Errorf("no VolumeGroupSnapshotClass found for driver %q for PVC %s", driver, pvc.Name)
+	} else {
+		return "", errors.Errorf("multiple VolumeGroupSnapshotClasses found for driver %q with label velero.io/csi-volumegroupsnapshot-class=true", driver)
+	}
+}
+
+func (p *pvcBackupItemAction) createVolumeGroupSnapshot(
+	backup *velerov1api.Backup,
+	pvc corev1api.PersistentVolumeClaim,
+	vgsLabelKey, vgsLabelValue, vgsClassName string,
+) (*volumegroupsnapshotv1beta1.VolumeGroupSnapshot, error) {
+	vgsLabels := map[string]string{
+		velerov1api.BackupNameLabel: label.GetValidName(backup.Name),
+		velerov1api.BackupUIDLabel:  string(backup.UID),
+		vgsLabelKey:                 vgsLabelValue,
+	}
+
+	vgs := &volumegroupsnapshotv1beta1.VolumeGroupSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("velero-%s-", vgsLabelValue),
+			Namespace:    pvc.Namespace,
+			Labels:       vgsLabels,
+		},
+		Spec: volumegroupsnapshotv1beta1.VolumeGroupSnapshotSpec{
+			VolumeGroupSnapshotClassName: &vgsClassName,
+			Source: volumegroupsnapshotv1beta1.VolumeGroupSnapshotSource{
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						vgsLabelKey: vgsLabelValue,
+					},
+				},
+			},
+		},
+	}
+
+	if err := p.crClient.Create(context.TODO(), vgs); err != nil {
+		return nil, errors.Wrap(err, "failed to create VolumeGroupSnapshot")
+	}
+
+	refetchedVGS, err := p.getVGSByLabels(pvc.Namespace, vgsLabels)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to re-fetch VGS after creation")
+	}
+
+	p.log.Infof("Re-fetched Created VolumeGroupSnapshot %s/%s for PVC group label %s=%s",
+		refetchedVGS.Namespace, refetchedVGS.Name, vgsLabelKey, vgsLabelValue)
+
+	return refetchedVGS, nil
+}
+
+func (p *pvcBackupItemAction) waitForVGSAssociatedVS(ctx context.Context, vgs *volumegroupsnapshotv1beta1.VolumeGroupSnapshot, timeout time.Duration) error {
+	return wait.PollUntilContextTimeout(ctx, time.Second, timeout, true, func(ctx context.Context) (done bool, err error) {
+		vsList := &snapshotv1api.VolumeSnapshotList{}
+		if err := p.crClient.List(ctx, vsList, crclient.InNamespace(vgs.Namespace)); err != nil {
+			return false, err
+		}
+
+		expected := 0
+		reported := 0
+
+		for _, vs := range vsList.Items {
```

Comment:
In most cases, it's safe to calculate the expected number of VSs in the VGS by filtering VSs with the owner reference.
It's still possible that some expected VSs are not created, and the created ones are ready.
It's better to calculate the expected number by the labelled PVCs.

Reply:
Vote on it. The expected number should not dynamically calculated. But I am not sure if we need to calculate it according to the number of PVC or there is any better solution.

Reply:
It is also not ideal to list the VS in a polling function, considering the performance in a scaled case.

Reply:
Updated the PR.